### ISSUE TITLE
edge.use returns 'road' instead of 'footways' for pedestrian costing

### ISSUE THREAD

**Issue by ZhaohanJackWang** [valhalla/valhalla#3561]
Hi there,

I'm trying to use the map matching feature of Valhalla to snap coordinates to footpathes and pedestrian crossings. To achieve this, I have specified the costing mode to be pedestrian. However, some coordinates were snapped to roads instead of footways. I tried to adjust some parameters but they didn't seem to work. I am wondering if there exists any solution to this. Thanks in advance.

**Comment by nilsnolde** [comment]
Do you have a minimal example request (ie ideally a short route)?

**Comment by ZhaohanJackWang** [comment]
Hi Nils,

Thanks for your reply. I've created a short request using: 
`'{"shape":[{"time":1633923297000,"lat":-37.813927,"lon":144.95715},{"time":1633923334000,"lat":-37.81397,"lon":144.95717},{"time":1633923342000,"lat":-37.814007,"lon":144.95717},{"time":1633923348000,"lat":-37.814045,"lon":144.9572},{"time":1633923352000,"lat":-37.814083,"lon":144.95718},{"time":1633923357000,"lat":-37.814133,"lon":144.95721},{"time":1633923363000,"lat":-37.814194,"lon":144.9572},{"time":1633923369000,"lat":-37.81426,"lon":144.9572},{"time":1633923374000,"lat":-37.814323,"lon":144.95723},{"time":1633923378000,"lat":-37.814384,"lon":144.95724},{"time":1633923384000,"lat":-37.814453,"lon":144.95728},{"time":1633923388000,"lat":-37.81452,"lon":144.95729},{"time":1633923393000,"lat":-37.81457,"lon":144.95734},{"time":1633923400000,"lat":-37.814632,"lon":144.95734},{"time":1633923406000,"lat":-37.814697,"lon":144.95732},{"time":1633923407000,"lat":-37.814686,"lon":144.9574},{"time":1633923409000,"lat":-37.814686,"lon":144.95749},{"time":1633923412000,"lat":-37.814747,"lon":144.95752},{"time":1633923415000,"lat":-37.81478,"lon":144.95755},{"time":1633923420000,"lat":-37.814835,"lon":144.9576}], "shape_match":"map_snap", "costing":"pedestrian",     "costing_options":{"pedestrian":{"walking_speed":"4.1"}},     "trace_options":{"turn_penalty_factor":500, "search_radius": 50, "gps_accuracy": 5},     "filters":{"attributes":["edge.travel_mode", "edge.way_id", "edge.use", "matched.point", "matched.edge_index"], "action":"include"}}'`

The edges I got:
`[{'use': 'road', 'travel_mode': 'pedestrian', 'way_id': 13366394},
 {'use': 'road', 'travel_mode': 'pedestrian', 'way_id': 13366394}]`

And, the matched_points:
`[{'edge_index': 0, 'lat': -37.813966, 'lon': 144.957019},
 {'edge_index': 0, 'lat': -37.814009, 'lon': 144.957039},
 {'edge_index': 0, 'lat': -37.814041, 'lon': 144.957055},
 {'edge_index': 0, 'lat': -37.814082, 'lon': 144.957074},
 {'edge_index': 0, 'lat': -37.81411, 'lon': 144.957088},
 {'edge_index': 0, 'lat': -37.814162, 'lon': 144.957112},
 {'edge_index': 0, 'lat': -37.814213, 'lon': 144.957136},
 {'edge_index': 0, 'lat': -37.814271, 'lon': 144.957164},
 {'edge_index': 0, 'lat': -37.814334, 'lon': 144.957193},
 {'edge_index': 0, 'lat': -37.81439, 'lon': 144.95722},
 {'edge_index': 0, 'lat': -37.814461, 'lon': 144.957254},
 {'edge_index': 0, 'lat': -37.814522, 'lon': 144.957283},
 {'edge_index': 0, 'lat': -37.814579, 'lon': 144.95731},
 {'edge_index': 0, 'lat': -37.814633, 'lon': 144.957335},
 {'edge_index': 0, 'lat': -37.814685, 'lon': 144.95736},
 {'edge_index': 0, 'lat': -37.814696, 'lon': 144.957365},
 {'edge_index': 0, 'lat': -37.81472, 'lon': 144.957376},
 {'edge_index': 18446744073709551615, 'lat': -37.814781, 'lon': 144.957406},
 {'edge_index': 1, 'lat': -37.81479, 'lon': 144.95741},
 {'edge_index': 1, 'lat': -37.81479, 'lon': 144.95741}]`

**Comment by nilsnolde** [comment]
thanks, that's a very short route :)

actually it makes sense, that way is a `highway=service`: https://www.openstreetmap.org/way/13366394, so it's generally open for all modes of transport, since there's nothing pedestrian-specific about it. generally people would want to avoid walking on service roads (usually used for parking aisles, access driveways for supermarkets, drive-ins etc), so we have the option to penalize using ways like that: https://github.com/valhalla/valhalla/blob/master/docs/api/turn-by-turn/api-reference.md#pedestrian-costing-options

in short, `use` is classifying the most prevalent property about the way, see more cases here: https://github.com/valhalla/valhalla/blob/master/proto/trip.proto#L17 (though some of them might share the same `use` value in the response)

**Comment by ZhaohanJackWang** [comment]
Hi Nils,

Sorry I'm still a bit confused. For this route:
`'{"shape":[{"time":1633923663000,"lat":-37.81761,"lon":144.9589},{"time":1633923666000,"lat":-37.817673,"lon":144.95885},{"time":1633923671000,"lat":-37.81773,"lon":144.95886},{"time":1633923678000,"lat":-37.817764,"lon":144.9589},{"time":1633923683000,"lat":-37.817787,"lon":144.95892},{"time":1633923687000,"lat":-37.81783,"lon":144.95895},{"time":1633923692000,"lat":-37.81788,"lon":144.95898},{"time":1633923697000,"lat":-37.81793,"lon":144.95905},{"time":1633923700000,"lat":-37.81795,"lon":144.95912},{"time":1633923704000,"lat":-37.81796,"lon":144.9592},{"time":1633923706000,"lat":-37.81799,"lon":144.95926},{"time":1633923709000,"lat":-37.818016,"lon":144.95932},{"time":1633923712000,"lat":-37.818058,"lon":144.95937},{"time":1633923715000,"lat":-37.818058,"lon":144.95943},{"time":1633923718000,"lat":-37.818077,"lon":144.95949},{"time":1633923720000,"lat":-37.818092,"lon":144.95952},{"time":1633923723000,"lat":-37.818115,"lon":144.95956},{"time":1633923727000,"lat":-37.818157,"lon":144.95961},{"time":1633923731000,"lat":-37.818222,"lon":144.95964},{"time":1633923736000,"lat":-37.818283,"lon":144.95969},{"time":1633923741000,"lat":-37.818333,"lon":144.95973},{"time":1633923745000,"lat":-37.818394,"lon":144.95976},{"time":1633923750000,"lat":-37.818443,"lon":144.9598},{"time":1633923755000,"lat":-37.81849,"lon":144.95976},{"time":1633923756000,"lat":-37.81854,"lon":144.95969},{"time":1633923757000,"lat":-37.81857,"lon":144.95963},{"time":1633923759000,"lat":-37.818596,"lon":144.95953},{"time":1633923761000,"lat":-37.818615,"lon":144.95947},{"time":1633923765000,"lat":-37.81863,"lon":144.95952},{"time":1633923769000,"lat":-37.818665,"lon":144.95956},{"time":1633923772000,"lat":-37.81871,"lon":144.9596},{"time":1633923776000,"lat":-37.81877,"lon":144.9597},{"time":1633923784000,"lat":-37.81884,"lon":144.95973},{"time":1633923791000,"lat":-37.818905,"lon":144.9598},{"time":1633923798000,"lat":-37.818974,"lon":144.95985},{"time":1633923803000,"lat":-37.819035,"lon":144.95984},{"time":1633923807000,"lat":-37.819084,"lon":144.95981},{"time":1633923809000,"lat":-37.81915,"lon":144.95976},{"time":1633923811000,"lat":-37.8192,"lon":144.9597},{"time":1633923818000,"lat":-37.819252,"lon":144.95978},{"time":1633923824000,"lat":-37.819298,"lon":144.95984},{"time":1633923830000,"lat":-37.819366,"lon":144.95988},{"time":1633923834000,"lat":-37.819427,"lon":144.95992},{"time":1633923836000,"lat":-37.81947,"lon":144.95987},{"time":1633923839000,"lat":-37.81952,"lon":144.95981},{"time":1633923842000,"lat":-37.81957,"lon":144.95978},{"time":1633923845000,"lat":-37.819603,"lon":144.95976},{"time":1633923849000,"lat":-37.81965,"lon":144.95975},{"time":1633923852000,"lat":-37.819687,"lon":144.95973},{"time":1633923855000,"lat":-37.81973,"lon":144.95973},{"time":1633923866000,"lat":-37.819683,"lon":144.95973},{"time":1633923873000,"lat":-37.819637,"lon":144.95972},{"time":1633923879000,"lat":-37.819603,"lon":144.95972}], "shape_match":"map_snap", "costing":"pedestrian",     "costing_options":{"pedestrian":{"driveway_factor":"10", "walkway_factor":"1"}},     "trace_options":{"turn_penalty_factor":500, "search_radius": 50, "gps_accuracy": 5},     "filters":{"attributes":["edge.travel_mode", "edge.way_id", "edge.use", "matched.point", "matched.edge_index"], "action":"include"}}'`

Some points were snapped to a primary road (`{'use': 'road', 'travel_mode': 'pedestrian', 'way_id': 878771432}`), which is just adjacent to a sidewalk. I also tried to increase the driveway_factor and reduce the walkway_factor using `"costing_options":{"pedestrian":{"driveway_factor":"10", "walkway_factor":"1"}}`. However, the results seem to be unchanged.

**Comment by nilsnolde** [comment]
hm, I can't reproduce what you see with https://valhalla1.openstreetmap.de: the way you're referencing is not included in my response. for me it starts on the other side of the street (way 505256329), then goes south to the pedestrian crossing at 712187002 to go on the footway on 658617482. that's what I'd expect as well (except the very first way also has a footway next to it which it somehow doesn't take):

<details>
<summary>response</summary>
<pre>
{
  "matched_points": [
    {
      "edge_index": 0,
      "lat": -37.817562,
      "lon": 144.959165
    },
    {
      "edge_index": 18446744073709551615,
      "lat": -37.817651,
      "lon": 144.959199
    },
    {
      "edge_index": 18446744073709551615,
      "lat": -37.817651,
      "lon": 144.959199
    },
    {
      "edge_index": 18446744073709551615,
      "lat": -37.817738,
      "lon": 144.959239
    },
    {
      "edge_index": 18446744073709551615,
      "lat": -37.817738,
      "lon": 144.959239
    },
    {
      "edge_index": 2,
      "lat": -37.817747,
      "lon": 144.959243
    },
    {
      "edge_index": 2,
      "lat": -37.817798,
      "lon": 144.959267
    },
    {
      "edge_index": 2,
      "lat": -37.817860,
      "lon": 144.959295
    },
    {
      "edge_index": 2,
      "lat": -37.817896,
      "lon": 144.959311
    },
    {
      "edge_index": 2,
      "lat": -37.817925,
      "lon": 144.959324
    },
    {
      "edge_index": 2,
      "lat": -37.817966,
      "lon": 144.959343
    },
    {
      "edge_index": 2,
      "lat": -37.818004,
      "lon": 144.959361
    },
    {
      "edge_index": 2,
      "lat": -37.818053,
      "lon": 144.959385
    },
    {
      "edge_index": 2,
      "lat": -37.818070,
      "lon": 144.959393
    },
    {
      "edge_index": 2,
      "lat": -37.818102,
      "lon": 144.959410
    },
    {
      "edge_index": 2,
      "lat": -37.818122,
      "lon": 144.959419
    },
    {
      "edge_index": 2,
      "lat": -37.818152,
      "lon": 144.959434
    },
    {
      "edge_index": 2,
      "lat": -37.818202,
      "lon": 144.959457
    },
    {
      "edge_index": 2,
      "lat": -37.818267,
      "lon": 144.959488
    },
    {
      "edge_index": 2,
      "lat": -37.818333,
      "lon": 144.959519
    },
    {
      "edge_index": 2,
      "lat": -37.818388,
      "lon": 144.959545
    },
    {
      "edge_index": 2,
      "lat": -37.818449,
      "lon": 144.959574
    },
    {
      "edge_index": 18446744073709551615,
      "lat": -37.818509,
      "lon": 144.959603
    },
    {
      "edge_index": 18446744073709551615,
      "lat": -37.818535,
      "lon": 144.959617
    },
    {
      "edge_index": 18446744073709551615,
      "lat": -37.818559,
      "lon": 144.959629
    },
    {
      "edge_index": 3,
      "lat": -37.818569,
      "lon": 144.959634
    },
    {
      "edge_index": 18446744073709551615,
      "lat": -37.818610,
      "lon": 144.959653
    },
    {
      "edge_index": 18446744073709551615,
      "lat": -37.818610,
      "lon": 144.959653
    },
    {
      "edge_index": 18446744073709551615,
      "lat": -37.818636,
      "lon": 144.959665
    },
    {
      "edge_index": 18446744073709551615,
      "lat": -37.818687,
      "lon": 144.959691
    },
    {
      "edge_index": 18446744073709551615,
      "lat": -37.818687,
      "lon": 144.959691
    },
    {
      "edge_index": 6,
      "lat": -37.818761,
      "lon": 144.959729
    },
    {
      "edge_index": 6,
      "lat": -37.818829,
      "lon": 144.959763
    },
    {
      "edge_index": 6,
      "lat": -37.818905,
      "lon": 144.959802
    },
    {
      "edge_index": 6,
      "lat": -37.818978,
      "lon": 144.959839
    },
    {
      "edge_index": 6,
      "lat": -37.819027,
      "lon": 144.959864
    },
    {
      "edge_index": 6,
      "lat": -37.819061,
      "lon": 144.959881
    },
    {
      "edge_index": 7,
      "lat": -37.819124,
      "lon": 144.959913
    },
    {
      "edge_index": 7,
      "lat": -37.819135,
      "lon": 144.959918
    },
    {
      "edge_index": 7,
      "lat": -37.819201,
      "lon": 144.959950
    },
    {
      "edge_index": 7,
      "lat": -37.819257,
      "lon": 144.959977
    },
    {
      "edge_index": 7,
      "lat": -37.819327,
      "lon": 144.960010
    },
    {
      "edge_index": 18446744073709551615,
      "lat": -37.819391,
      "lon": 144.960041
    },
    {
      "edge_index": 18446744073709551615,
      "lat": -37.819507,
      "lon": 144.959891
    },
    {
      "edge_index": 18446744073709551615,
      "lat": -37.819513,
      "lon": 144.959874
    },
    {
      "edge_index": 10,
      "lat": -37.819547,
      "lon": 144.959769
    },
    {
      "edge_index": 10,
      "lat": -37.819556,
      "lon": 144.959738
    },
    {
      "edge_index": 10,
      "lat": -37.819564,
      "lon": 144.959710
    },
    {
      "edge_index": 10,
      "lat": -37.819566,
      "lon": 144.959703
    },
    {
      "edge_index": 10,
      "lat": -37.819566,
      "lon": 144.959703
    },
    {
      "edge_index": 10,
      "lat": -37.819566,
      "lon": 144.959703
    },
    {
      "edge_index": 10,
      "lat": -37.819566,
      "lon": 144.959703
    },
    {
      "edge_index": 10,
      "lat": -37.819566,
      "lon": 144.959703
    }
  ],
  "edges": [
    {
      "use": "road",
      "travel_mode": "pedestrian",
      "way_id": 505256329
    },
    {
      "use": "road",
      "travel_mode": "pedestrian",
      "way_id": 208642575
    },
    {
      "use": "road",
      "travel_mode": "pedestrian",
      "way_id": 208642575
    },
    {
      "use": "road",
      "travel_mode": "pedestrian",
      "way_id": 208642575
    },
    {
      "use": "road",
      "travel_mode": "pedestrian",
      "way_id": 208642575
    },
    {
      "use": "road",
      "travel_mode": "pedestrian",
      "way_id": 968051653
    },
    {
      "use": "road",
      "travel_mode": "pedestrian",
      "way_id": 968051653
    },
    {
      "use": "road",
      "travel_mode": "pedestrian",
      "way_id": 968051653
    },
    {
      "use": "pedestrian_crossing",
      "travel_mode": "pedestrian",
      "way_id": 712187002
    },
    {
      "use": "pedestrian_crossing",
      "travel_mode": "pedestrian",
      "way_id": 712187002
    },
    {
      "use": "footway",
      "travel_mode": "pedestrian",
      "way_id": 658617482
    }
  ],
  "alternate_paths": []
}
</pre>
</details>

**Comment by nilsnolde** [comment]
I'm not sure exactly, like I said, not really a map-matching user myself. I guess you could try some of the extra location properties, like `search_filter.max_road_class` and then also jacking up (and down) the factors which are relevant (as you already did, but probably best to use values < 1 to prefer ways).

**Comment by ZhaohanJackWang** [comment]
Ok, thank you for your help. I really appreciate it.
